File: Baekjun/Old Codes/1012/1012.cpp
#include <iostream>
#include <queue>
#include <cstring>
#define Max 50

int map[Max][Max];
int visited[Max][Max];

int T, M, N, K;

int dx[4] = { -1, 1, 0, 0 };
int dy[4] = { 0, 0, -1, 1 };

int bugCounter = 0;

using namespace std;

queue<pair<int, int>> q;

void printing()
{
        cout << endl;
    
    for (int i = 0; i < N; i++)
    {

        for (int j = 0;  j < M; j++)
        {
            cout << map[i][j] << " ";
        }
        cout << endl;
    }
}

void bfs(int col, int row)
{
    q.push(make_pair(col, row));
    visited[col][row] = 1;

    while(!q.empty())
    {
    

        int y = q.front().first;
        int x = q.front().second;

        q.pop();

        for (int i = 0; i < 4; i++)
        {
            int ny = y + dy[i];
            int nx = x + dx[i];
                    // cout << "[1] ny : " << ny << " nx : " << nx << endl;


            if( nx <0 || nx >=M || ny < 0 || ny >= N)
            {

                continue;
            }
            else{

                if(map[ny][nx] == 1 && visited[ny][nx] == 0)
                {
                    q.push(make_pair(ny, nx));
                    visited[ny][nx] = 1;
                    // cout << " ny : " << ny << " nx : " << nx << endl;

                }

            }


        }
    }

    bugCounter++;
}


void solve()
{

    for (int i = 0; i < N; i++)
    {

        for (int j = 0;  j < M; j++)
        {

            if(map[i][j] == 1 && visited[i][j] == 0)
            {
                // cout << " i : " << i << " j : " << j << endl;
                // cout << " map : " << map[i][j] << " visited : " << visited[i][j] << endl;
                bfs(i, j);

            }

        }
    }

    cout << bugCounter << endl;
}



void input()
{
    cin >> T;

    for (int i = 0; i < T; i ++)
    {

        memset(map, 0, sizeof(map));
        memset(visited, 0, sizeof(visited));
        bugCounter = 0;

        cin >> M >> N >> K;

        for (int j = 0; j < K; j++)
        {
            int tempX, tempY;
            cin >> tempX >> tempY;

            map[tempY][tempX] = 1;

        }

        solve();

        // printing();
        
    }

}





int main()
{

    input();




    return 0;
}

File: Algorithm_python/Baekjun/Category/Graph Theory/BFS/1012.py
from pprint import pprint

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def BFS(y, x):
    queue = [(y, x)]
    visited[y][x] = 1
    pprint(visited)

    while queue:
        cur_y, cur_x = queue.pop(0)
        print(cur_y, cur_x)

        for i in range(4):
            nx = cur_x + dx[i]
            ny = cur_y + dy[i]

            if nx < 0 or nx >= row or ny < 0 or ny >= col:
                continue
            if visited[ny][nx] == 1:
                # print("visited")
                continue
            if matrix[ny][nx] == 0:
                # print("matrix = 0")
                continue

            queue.append((ny, nx))
            visited[ny][nx] = 1
            # pprint(visited)

    print("BFS 종료")

T = int(input())

for _ in range(T):
    ans = 0
    row, col, K = map(int, input().split())

    matrix = [[0] * row for _ in range(col)]
    visited = [[0] * row for _ in range(col)]

    for _ in range(K):
        x, y = map(int, input().split())
        matrix[y][x] = 1

    pprint(matrix)

    for y in range(col):
        for x in range(row):
            if visited[y][x] != 1 and matrix[y][x] == 1:
                BFS(y, x)
                ans += 1
                # print("x: ", x)
                # print("y: ", y)
                # print("ans: ", ans)

    print(ans)